pkg/frontend/querymiddleware/astmapper/instant_splitting.go

```diff
@@ -0,0 +1,542 @@
+		splitOffset += originalOffset
+		splitNode, err := createSplitNode(embeddedQuery, splitRangeInterval, splitOffset)
+		if err != nil {
+			return nil, false, err
+		}
+
+		// Prepend to embedded queries
+		embeddedQueries = append([]parser.Node{splitNode}, embeddedQueries...)
+	}
+
+	squashExpr, err := vectorSquasher(embeddedQueries...)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return squashExpr, true, nil
+}
+
+// getRangeInterval returns the range interval in the range vector node
+// Returns 0 if no range interval is found
+// Example: expression `count_over_time({app="foo"}[10m])` returns 10m
+func getRangeInterval(node parser.Node) time.Duration {
+	switch n := node.(type) {
+	case *parser.AggregateExpr:
+		return getRangeInterval(n.Expr)
+	case *parser.Call:
+		argRangeInterval := time.Duration(0)
+		// Iterate over Call's arguments until a MatrixSelector is found
+		for _, arg := range n.Args {
+			if argRangeInterval = getRangeInterval(arg); argRangeInterval != 0 {
+				break
+			}
+		}
+		return argRangeInterval
+	case *parser.MatrixSelector:
+		return n.Range
+	default:
+		return 0
+	}
+}
+
+// getOffset returns the offset interval in the range vector node
+// Returns 0 if no offset interval is found
+// Example: expression `count_over_time({app="foo"}[10m]) offset 1m` returns 1m
+func getOffset(node parser.Node) time.Duration {
+	switch n := node.(type) {
+	case *parser.AggregateExpr:
+		return getOffset(n.Expr)
+	case *parser.Call:
+		argRangeInterval := time.Duration(0)
+		// Iterate over Call's arguments until a MatrixSelector is found
+		for _, arg := range n.Args {
+			if argRangeInterval = getOffset(arg); argRangeInterval != 0 {
+				break
+			}
+		}
+		return argRangeInterval
+	case *parser.MatrixSelector:
+		return getOffset(n.VectorSelector)
+	case *parser.VectorSelector:
+		return n.OriginalOffset
+	default:
+		return 0
+	}
+}
+
+// expr can be a parser.Call or a parser.AggregateExpr
+func createSplitNode(expr parser.Expr, rangeInterval time.Duration, offset time.Duration) (parser.Node, error) {
+	splitNode, err := cloneNode(expr)
+	if err != nil {
+		return nil, err
+	}
+	rangeIntervalUpdated := updateRangeInterval(splitNode, rangeInterval)
+	if !rangeIntervalUpdated {
+		return nil, fmt.Errorf("unable to update range interval on node: %v", splitNode)
+	}
+	offsetUpdated := updateOffset(splitNode, offset)
+	if !offsetUpdated {
+		return nil, fmt.Errorf("unable to update offset operator on node: %v", splitNode)
+	}
+
+	return splitNode, nil
+}
+
+// updateEmbeddedExpr returns the updated node if inner call expression was updated successfully,
+// otherwise returns nil
+func updateEmbeddedExpr(node parser.Node, call *parser.Call) parser.Node {
+	switch n := node.(type) {
+	case *parser.AggregateExpr:
+		embeddedNode := updateEmbeddedExpr(n.Expr, call)
+		embeddedExpr, ok := embeddedNode.(parser.Expr)
+		if !ok {
+			return nil
+		}
+		n.Expr = embeddedExpr
+		return n
+	case *parser.Call:
+		return call
```

Comment:
According to the function, the contract is that it returns `nil` if the node was not updated. Here it's not updated, why we return `call`?

Reply:
We need to return call, because of this assignment: https://github.com/grafana/mimir/blob/b84a2001af28cf4f31550d400eacac7e00695b65/pkg/frontend/querymiddleware/astmapper/instant_splitting.go#L473

I also don't like the way this function is implemented, but I think after rebasing @colega PR: https://github.com/grafana/mimir/pull/2556 this should be clearer